crates/matrix-sdk-crypto/src/backups/keys/decryption.rs

```diff
@@ -219,18 +242,39 @@ impl BackupDecryptionKey {
     /// [`BackupDecryptionKey`].
     pub fn decrypt_session_data(
         &self,
+        user_id: &UserId,
         session_data: EncryptedSessionData,
     ) -> Result<BackedUpRoomKey, DecryptionError> {
-        let message = Message {
-            ciphertext: session_data.ciphertext.into_inner(),
-            mac: session_data.mac.into_inner(),
-            ephemeral_key: Curve25519PublicKey::from_slice(session_data.ephemeral.as_bytes())?,
+        let message = match (&session_data.mac, &session_data.signatures) {
+            (_, Some(_)) => {
+                let mac_key = self.calculate_mac_key();
+                mac_key.verify(user_id, &DeviceKeyId::from_parts(DeviceKeyAlgorithm::HmacSha256, "backup_mac_key".into()), &session_data)
+                    .map_err(|_| DecryptionError::Mac(MacError))?;
+                Message {
+                    ciphertext: session_data.ciphertext.into_inner(),
+                    mac: None,
+                    ephemeral_key: Curve25519PublicKey::from_slice(session_data.ephemeral.as_bytes())?,
+                }
+            },
+            (Some(mac), _) => Message {
+                // FIXME: if we're in a V2 backup, this should be an error
```

Comment:
How do we handle V2 semantics?  Do we create a `BackupV2DecryptionKey` struct that gets used for V2 backups?  Is there an easy way t share code between `BackupDecryptionKey` and `BackupV2DecryptionKey`?